dbms/src/Disks/DiskLocal.cpp

```diff
@@ -0,0 +1,240 @@
+}
+
+bool DiskLocal::tryReserve(UInt64 bytes) const
+{
+    std::lock_guard lock(mutex);
+    if (bytes == 0)
+    {
+        LOG_DEBUG(&Logger::get("DiskLocal"), "Reserving 0 bytes on disk " << backQuote(name));
+        ++reservation_count;
+        return true;
+    }
+
+    auto available_space = getAvailableSpace();
+    UInt64 unreserved_space = available_space - std::min(available_space, reserved_bytes);
+    if (unreserved_space >= bytes)
+    {
+        LOG_DEBUG(
+            &Logger::get("DiskLocal"),
+            "Reserving " << formatReadableSizeWithBinarySuffix(bytes) << " on disk " << backQuote(name) << ", having unreserved "
+                         << formatReadableSizeWithBinarySuffix(unreserved_space) << ".");
+        ++reservation_count;
+        reserved_bytes += bytes;
+        return true;
+    }
+    return false;
+}
+
+UInt64 DiskLocal::getTotalSpace() const
+{
+    auto fs = getStatVFS(path);
+    UInt64 total_size = fs.f_blocks * fs.f_bsize;
+    if (total_size < keep_free_space_bytes)
+        return 0;
+    return total_size - keep_free_space_bytes;
+}
+
+UInt64 DiskLocal::getAvailableSpace() const
+{
+    /// we use f_bavail, because part of b_free space is
+    /// available for superuser only and for system purposes
+    auto fs = getStatVFS(path);
+    UInt64 total_size = fs.f_bavail * fs.f_bsize;
+    if (total_size < keep_free_space_bytes)
+        return 0;
+    return total_size - keep_free_space_bytes;
+}
+
+UInt64 DiskLocal::getUnreservedSpace() const
+{
+    std::lock_guard lock(mutex);
+    auto available_space = getAvailableSpace();
+    available_space -= std::min(available_space, reserved_bytes);
+    return available_space;
+}
+
+DiskFilePtr DiskLocal::file(const String & path_) const
+{
+    return std::make_shared<DiskLocalFile>(std::static_pointer_cast<const DiskLocal>(shared_from_this()), path_);
+}
+
+
+DiskLocalFile::DiskLocalFile(const DiskPtr & disk_ptr_, const String & rel_path_)
+    : IDiskFile(disk_ptr_, rel_path_), file(disk_ptr->getPath() + rel_path)
+{
+}
+
+bool DiskLocalFile::exists() const
+{
+    return file.exists();
+}
+
+bool DiskLocalFile::isDirectory() const
+{
+    return file.isDirectory();
+}
+
+void DiskLocalFile::createDirectory()
+{
+    file.createDirectory();
+}
+
+void DiskLocalFile::createDirectories()
+{
+    file.createDirectories();
+}
+
+void DiskLocalFile::moveTo(const String & new_path)
+{
+    file.renameTo(disk_ptr->getPath() + new_path);
+}
+
+void DiskLocalFile::copyTo(const String & new_path)
+{
+    file.copyTo(disk_ptr->getPath() + new_path);
+}
+
+std::unique_ptr<ReadBuffer> DiskLocalFile::read() const
+{
+    return std::make_unique<ReadBufferFromFile>(file.path());
+}
+
+std::unique_ptr<WriteBuffer> DiskLocalFile::write()
+{
+    return std::make_unique<WriteBufferFromFile>(file.path());
+}
+
+
+DiskDirectoryIteratorImplPtr DiskLocalFile::iterateDirectory()
```

Comment:
I think independent function is better here.

Reply:
DiskLocalFile has been removed.